### PR TITLE
Check deck tree update name map

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Purpose / Description
I was thinking "in case of problem, we just have to construct the name map again once they are corrected".
I didn't fully grasp that name map was actually used in the deck manager and that `_checkDeckTree` uses the deck manager, E.g. to ensure parents are created. So I need to ensure that the name map stays consistent during the whole computation.

## Fixes
Should fix https://github.com/ankidroid/Anki-Android/issues/6636 

## Approach
When a deck is renamed (e.g. because blank, because duplicate) the information is directly given to the name map, so that the remaining of the code can find this name. 
Furthermore, we uses a map instead of a set, this way, when we find a duplicate, we also find the original deck and ensure both are in the name map (the original may have been erased when adding the duplicate)

## How Has This Been Tested?

Running on my phone. However, as it's a problem that should rarely occurs, I can't test it in real life.

## Learning (optional, can help others)
Code which deals with accident (duplicate name, empty names)... is harder to think through than normal code, since usual assumptions are broken. Those assumption may be repaired, but the method which repair them still need to deal with the broken collection

@mikehardy I didn't pay attention when you added `isValidDeckName`, but the name is not really correct.
Nowaday, valid deck name should not contains "::::", and no part of the name should starts or end with spaces.  However, I don't know whether you want to send error message for spaces as we can easily trim them. That should occur in a different PR I assume

Deck names which are not normalized UTF-8 are not valid either, and I don't know if it's possible for a user to actually enter that except if they are really trying to do it

## Checklist
_Please, go through these checks before submitting the PR._

- [ ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [ ] You have a descriptive commit message with a short title (first line, max 50 chars).
- [ ] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ ] You have commented your code, particularly in hard-to-understand areas
- [ ] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.java [hunk 1]
```diff
@@ -885,44 +885,48 @@ private void _checkDeckTree() {
...
 
             // two decks with the same name?
-            if (names.containsKey(normalizeName(deck.getString("name")))) {
-                Timber.i("fix duplicate deck name %s", deck.getString("name"));
+            if (names.containsKey(normalizeName(deckName))) {
+                Timber.i("fix duplicate deck name %s", deckName);
                 do {
-                    deck.put("name", deck.getString("name") + "+");
-                } while (names.containsKey(normalizeName(deck.getString("name"))));
+                    deckName = deckName + "+";
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Decks.java:909]:
```suggestion
                    deckName += "+";
```

**Comment by Arthur-Milchior** [comment]
I added a test. 
I believe the test should create a manual exception before and not after the PR
I don't know how to test it, since we are not actually raising exception and I expect that the method actually works before as it work after

**Comment by Arthur-Milchior** [comment]
one of the main problem I have here is that the bug will or won't occur depending on the order in which HashMap.values() returns its elements. There is one chance out of two that the element renamed from "A" to "A+" is the one in the name map. I don't know how to ensure that it is always the case

Because of this, even if I replace the `AnkiDroidApp.sendExceptionReport` by a real exception, I can't get it.

**Comment by mikehardy** [comment]
isValidDeckName is definitely *intended* to be a deck name check prior to creating a deck, but I think in our new naming we should call it "isValidDeckNameComponent" (since it's for the UI version of a deck, which is actually just a part of a potential deck hierarchy where the whole name is encoded with the :: separators yes)

But I certainly don't know all the validation rules, all I know is that they were not validated *at all* before, so I put a first take in place where I at least don't allow completely empty names. Any expansion is welcome but is also not necessarily a priority - we lived without any validation this long after all...

If there is difficult testing based on ordering going on, you can swap in a different Collection right? I just had this problem recently in another project and swapped in a TreeMap to get predictable ordering vs HashMap. To make it "testable" you can extract the part where you `new HashMap` to maybe a "getMap" and then make a TestDecks object in the test tree that overrides getMap and returns your predictably ordered one, or similar. I use little tricks like that frequently, you can see in the tests directory any object named xxxxTest.java contains tests, and all the TestXxxx.java objects are little testability sub-classes where I swap things in that are test-specific

Sometimes it is very tricky to think about how to make something testable, but the answer is frequently just 2 little 5-line chunks of code

**Comment by mikehardy** [comment]
marking as pending feedback to see if the idea of a test subclass that imrpoves the test is quickly possible. might be? might not be? hard to say until you try pushing the code around in an attempt

**Comment by Arthur-Milchior** [comment]
The trouble I have is that my test class never see the collection. IT's not like it gets the list at any point. The only contents it has is some JSON string representing a set of deck. The hash map is created by the deck manager, an iterator is exctracted from it by HashMap class, and this iterator is used in `safetyCheck`. At no point in time does the test has the list. So, unless I just copy/paste some part of the code in the test class, there is nothing I can do here on the list

**Comment by mikehardy** [comment]
It's possible either in the main class to mark things `@VisibleForTesting` so you can expose access to the collection or to override and have Test object with accessors. In Testing anything really is possible for inspection. I'll merge this because I think it's really important but whatever you think you need access to in a Test, you can get access to - you shouldn't mutate internal state in tests (it would defeat the purpose) but cracking open objects in order to inspect them and assert on state? Anything goes I think, and there's lots of tricks for it

**Pull request merged**